tests/std/tests/P0768R1_spaceship_cpos/test.cpp

```diff
@@ -0,0 +1,554 @@
+static_assert(NoexceptCpo<compare_weak_order_fallback, SpaceshipType<NonThrowing>>); // [cmp.alg]/5.2
+static_assert(NoexceptCpo<compare_partial_order_fallback, SpaceshipType<NonThrowing>>); // [cmp.alg]/6.2
+
+// Test fallbacks. Part B, exception specifications.
+static_assert(NoexceptCpo<compare_strong_order_fallback, Fallback<0>>); // [cmp.alg]/4.3
+static_assert(!NoexceptCpo<compare_strong_order_fallback, Fallback<1>>); // [cmp.alg]/4.3
+static_assert(!NoexceptCpo<compare_strong_order_fallback, Fallback<5>>); // [cmp.alg]/4.3
+
+static_assert(NoexceptCpo<compare_weak_order_fallback, Fallback<0>>); // [cmp.alg]/5.3
+static_assert(!NoexceptCpo<compare_weak_order_fallback, Fallback<1>>); // [cmp.alg]/5.3
+static_assert(!NoexceptCpo<compare_weak_order_fallback, Fallback<5>>); // [cmp.alg]/5.3
+
+static_assert(NoexceptCpo<compare_partial_order_fallback, Fallback<0>>); // [cmp.alg]/6.3
+static_assert(!NoexceptCpo<compare_partial_order_fallback, Fallback<1>>); // [cmp.alg]/6.3
+static_assert(!NoexceptCpo<compare_partial_order_fallback, Fallback<5>>); // [cmp.alg]/6.3
+
+static_assert(!NoexceptCpo<compare_partial_order_fallback, Fallback<10>, const Fallback<10>>); // [cmp.alg]/6.3
+
+// Define test machinery for compile-time and run-time behavior.
+template <const auto& CPO, typename T>
+constexpr void test_behavior() {
+    const T one{1};
+    const T two{2};
+
+    using Ordering = CpoResult<CPO, const T&>;
+
+    assert(CPO(one, two) == Ordering::less);
+    assert(CPO(two, two) == Ordering::equivalent);
+    assert(CPO(two, one) == Ordering::greater);
+}
+
+// This takes an array of pair<int, Floating>, where the ints are arbitrary "ranks" for the std::weak_order equivalence
+// classes. It tests std::strong_order, std::weak_order, and std::partial_order with every permutation of values.
+template <typename PairArray>
+constexpr void test_ranked_values(const PairArray& rank_value_pairs) {
+    constexpr size_t N = extent_v<PairArray>;
+
+    for (size_t l_idx = 0; l_idx < N; ++l_idx) {
+        const auto& [l_rank, left] = rank_value_pairs[l_idx];
+        for (size_t r_idx = 0; r_idx < N; ++r_idx) {
+            const auto& [r_rank, right] = rank_value_pairs[r_idx];
+
+            assert(std::strong_order(left, right) == l_idx <=> r_idx); // [cmp.alg]/1.3
+            assert(std::weak_order(left, right) == l_rank <=> r_rank); // [cmp.alg]/2.3
+            assert(std::partial_order(left, right) == left <=> right); // [cmp.alg]/3.3 (uses compare_three_way)
+        }
+    }
+}
+
+template <typename Floating>
+constexpr void test_floating() {
+    if constexpr (is_same_v<Floating, float>) {
+        const pair<int, float> rank_value_pairs[]{
+            {10, bit_cast<float>(0xFFC01234u)}, // negative quiet NaN, payload bits
+            {10, bit_cast<float>(0xFFC00000u)}, // negative quiet NaN, no payload bits
+            {20, -numeric_limits<float>::infinity()}, // negative infinity

```

Comment:
Do we need test coverage for signaling NaNs?

Note that it could be currently impossible to generate a `float` signaling NaN on the MSVC compiler at compile time. It emits a quiet NaN for `bit_cast<float>(0xFF800001u)` or `numeric_limits<float>::signaling_NaN()`.

Same below.

Reply:
Ideally, yes, this would test signaling NaNs - but as you noted, the compiler bugs make it difficult. I think we can live without test coverage for them, given that the code is clearly written to handle them uniformly (the main question is whether I got the "infinity plus one" logic correct, but I think that can be verified by inspecting the code).